doc/architectural/goto-program-transformations.md

```diff
@@ -0,0 +1,249 @@
+\ingroup module_hidden
+
+\page goto-program-transformations Goto Program Transformations
+
+TODO:
+- Add link for each pass to the function that implements it.
+- Add to each title the name of the function that implements it.
+- Add link points and cross-linking
+- Fix consistency of Removal/Remove etc.
+
+## Core Transformation Passes
+
+This section lists the transformation passes that must have been applied for the
+goto model to be in core goto format.
+
+Note that the passes are listed below in the order they are currently called in
+CBMC. While all dependencies on the ordering are not fully documented, the
+following order should be used.
+
+### Removal/Lowering of Assembly
+
+This pass goes through the goto model and removes or lowers instances of
+assembly intructions. Assembly instructions are stored in instances of the
+`other` instruction.
+
+_This pass has no predeccesor._
+
+### String Instrumentation
+
+This (optional) pass adds checks on calls to some of the C standard library
+string functions. It uses the tracking information added by the
+[string-abstraction] pass. It is switched on by the `--string-abstraction`
+command line option.
```

Comment:
There is a description of what it does associated with the `string_abstractiont` class. My understanding is that it only adds tracking information rather than discarding any information. So there are no properties which you can't analyse when the pass is in place, which you can analyse when it isn't. Tracking the length simply helps improve the efficiency of analysing length based assertions because it avoids having to iterate through C strings to work out how long they are as part of the assertion.